polys: Replace algorithm of primitive_element for ex=True

jksuom opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #20163

#### Brief description of what is fixed or changed


The algorithm currently used by primitive_element for ex=True is
based on the `groebner` function. That does not work reliably if
the input extension elements are not linearly disjoint (see e.g.
#14117). This commit implements the same iterative method
that is used for the ex=False case extended
by including representations of field elements.


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- polys
  - numberfields: implemented new algorithm for `primitive_element` in case `ex=True`.
<!-- END RELEASE NOTES -->

jksuom committed:
Make AlgebraicNumber alias symbol non-free.

minimal_polynomial fails on AlgebraicNumbers that have free symbols.

jksuom committed:
Replace algorithm of primitive_element for ex=True.

The algorithm currently used by primitive_element for ex=True is
based on the groebner function. That does not work reliably if
the input extension elements are not linearly disjoint (see e.g.
https://github.com/sympy/sympy/issues/14117). This commit implements
the same iterative method that is used for the ex=False case extended
by including representations of field elements.

jksuom committed:
Test for issue 20163.

jksuom committed:
Remove CoercionFailed and variations.

oscarbenjamin commented:
Let me try to understand how this works...

We want to construct an algebraic number field containing some expressions like:
```julia
In [4]: exts = [sqrt(3)+I, -sqrt(3)+I, sqrt(3)-I, -sqrt(3)-I, -2*I, 2*I]                                                                       

In [5]: construct_domain(exts, extension=True)                                                                                                 
^C---------------------------------------------------------------------------
KeyboardInterrupt
```
Clearly these extensions are all generated by `sqrt(3)` and `I` over `QQ`. In fact by the primitive element theorem we can have a single generator like `sqrt(3) + I` so it can be `QQ<sqrt(3) + I>` with minimal polynomial:
```julia
In [6]: minimal_polynomial(sqrt(3) + I)                                                                                                        
Out[6]: 
 4      2     
x  - 4⋅x  + 16
```
On master this primitive element is calculated by computing the minimal polynomials for each of `exts` and then computing their Groebner basis which can be very slow.

In this PR it is instead computed without the Groebner basis and succeeds in finding a primitive element `-sqrt(3)-I` (the primitive element found is non-deterministic):
```julia
In [1]: exts = [sqrt(3)+I, -sqrt(3)+I, sqrt(3)-I, -sqrt(3)-I, -2*I, 2*I]                                                                       

In [2]: construct_domain(exts, extension=True)                                                                                                 
Out[2]: 
(QQ<-sqrt(3) - I>,
 [ANP([mpq(-1,1), mpq(0,1)], [mpq(1,1), mpq(0,1), mpq(-4,1), mpq(0,1), mpq(16,1)], QQ),
  ANP([mpq(-1,4), mpq(0,1), mpq(1,1), mpq(0,1)], [mpq(1,1), mpq(0,1), mpq(-4,1), mpq(0,1), mpq(16,1)], QQ),
  ANP([mpq(1,4), mpq(0,1), mpq(-1,1), mpq(0,1)], [mpq(1,1), mpq(0,1), mpq(-4,1), mpq(0,1), mpq(16,1)], QQ),
  ANP([mpq(1,1), mpq(0,1)], [mpq(1,1), mpq(0,1), mpq(-4,1), mpq(0,1), mpq(16,1)], QQ),
  ANP([mpq(1,4), mpq(0,1), mpq(0,1), mpq(0,1)], [mpq(1,1), mpq(0,1), mpq(-4,1), mpq(0,1), mpq(16,1)], QQ),
  ANP([mpq(-1,4), mpq(0,1), mpq(0,1), mpq(0,1)], [mpq(1,1), mpq(0,1), mpq(-4,1), mpq(0,1), mpq(16,1)], QQ)])
```

The method used here receives a slightly filtered list of extensions:
```python
[I, sqrt(3) + I, sqrt(3) - I, -sqrt(3) - I, -sqrt(3) + I]
```
This has combined `2*I` and `-2*I` and removed the rational factor giving just `I`. I'm not sure where that preprocessing happened but would it not have been straight-forward to process the Adds as well to end up with just `[I, sqrt(3)]`?

In any case the method proceeds by computing the minimal polynomial for each element over `QQ` and then attempting to combine those into a single extension field. In the first iteration we take `QQ<exts[0]>` and `QQ<exts[1]>` and then try to factor one minimal polynomial over the other extension field. I guess if that succeeds then the minimal polynomial for one is reducible over the other field and so it's generators are not needed:
```julia
In [7]: ext1 = I                                                                                                                               

In [8]: ext2 = sqrt(3) + I                                                                                                                     

In [9]: f = minimal_polynomial(ext1, x, polys=True)                                                                                            

In [10]: K = QQ.algebraic_field((f, ext1))                                                                                                     

In [11]: g = minimal_polynomial(ext2, x, polys=True)                                                                                           

In [12]: L = QQ.algebraic_field((f, ext2))                                                                                                     

In [13]: f                                                                                                                                     
Out[13]: Poly(x**2 + 1, x, domain='QQ')

In [14]: g                                                                                                                                     
Out[14]: Poly(x**4 - 4*x**2 + 16, x, domain='QQ')

In [15]: K                                                                                                                                     
Out[15]: QQ<I>

In [16]: L                                                                                                                                     
Out[16]: QQ<sqrt(3) + I>

In [17]: factor_list(f, domain=L)                                                                                                              
Out[17]: (1, [(Poly(x - sqrt(3) - I, x, domain='QQ<sqrt(3) + I>'), 1), (Poly(x + sqrt(3) + I, x, domain='QQ<sqrt(3) + I>'), 1)])
```
I'm not quite sure what the `sqf_norm` and the `gcd` are doing but I guess they are trying to normalise the result in some way.

It seems that the loop will attempt to convert all reps to a new field on each iteration. Perhaps it should be checked if the new K is actually different from the previous K.

Why is the method non-deterministic? Maybe the ordering of `extension` comes from a `set` somewhere.

Is it possible to preprocess this and extract the unique non-rational generators from the Adds?

Presumably this does not fix #14177 so what are examples that still don't work? The OP example seems to work with this change.

jksuom commented:
This works by adjoining one `ext` element at a time to Q. The primitive element of the growing extension is denoted by `gen`. At each step an `ext` is adjoined to `K = Q(gen)`. `K(ext)` can then be generated by a rational linear combination `gen + s*ext` for all but a finite number of coefficients `s`. It is the job of `sqf_norm` to find a suitable `s`. It tries `s = 0, 1, 2, ...` and stops at the first success. (So, essentially, this is quite deterministic.)

For a polynomial `f` with domain `Q(ext)`, `f.sqf_norm()` returns `s, g, r` where `g(x) = f(x - s*ext)` (there is a typo in the docstring), `r` is its norm from `Q(ext)` to `Q` and `s` has been chosen such that `r` is squarefree. In the algorithm, `f` will be the minimal polynomial of `gen` over `Q(ext)`, hence irreducible. Then `g` will also be irreducible and its norm will be a power of an irreducible polynomial, so also irreducible being squarefree. Hence it will be a minimal polynomial of the new generator `gen + s*ext`.

> Perhaps it should be checked if the new K is actually different from the previous K.

This would be the case `s == 0`. It could be tested but I'm not sure that it would be worthwhile.
It will only happen with redundant `ext` elements.

The role of `gcd` may look obscure. The polynomial `g` is an algebraic relation of the new `gen` over `Q(ext)`. It is first transformed to a relation `h` of `ext` over `Q(gen)`. We know that there is a linear relation as `ext` is an element of `Q(gen)`. Taking the `gcd` of `h` with the minimal polynomial of `ext` over `Q` will create such a relation.

oscarbenjamin commented:
There is definitely something non-deterministic. Given `t.py`:
```python
from sympy import *

exts = [sqrt(3)+I, -sqrt(3)+I, sqrt(3)-I, -sqrt(3)-I, -2*I, 2*I]
print(construct_domain(exts, extension=True)[0])
```
I get different results on different runs:
```console
$ for n in `seq 1 10`; do python t.py ; done
QQ<sqrt(3) + I>
QQ<-2*sqrt(3) - I>
QQ<sqrt(3) - I>
QQ<-2*sqrt(3) - I>
QQ<-sqrt(3) - I>
QQ<-sqrt(3) - I>
QQ<-sqrt(3) + 2*I>
QQ<-sqrt(3) - I>
QQ<-sqrt(3) - I>
QQ<sqrt(3) + I>
```

oscarbenjamin commented:
> It will only happen with redundant `ext` elements.

In the example:
```
[I, sqrt(3) + I, sqrt(3) - I, -sqrt(3) - I, -sqrt(3) + I]
```
almost all of the elements are redundant. Presumably the common case for `construct_domain` is that we have a possibly large number of elements but all coming from the same extension field. Maybe there would be less redundancy if the list was prefiltered to `[I, sqrt(3)]`.

jksuom commented:
> Maybe there would be less redundancy if the list was prefiltered to [I, sqrt(3)].

That would certainly be true. The prefiltering should be done before calling `primitive_element`.

oscarbenjamin commented:
I think that it does make sense for `construct_domain` to break up the terms of an `Add` to find the elementary algebraic extension elements before calling `primitive_element`. Wouldn't it also be good for `primitive_element` to give a deterministic result regardless of the ordering of the inputs though?

For example on master we have:
```julia
In [1]: a = sqrt(2); b = 1 - sqrt(2)                                                                                                           

In [2]: primitive_element([a, b])                                                                                                              
Out[2]: 
⎛ 2            ⎞
⎝x  - 2, [1, 0]⎠

In [3]: primitive_element([a, b])                                                                                                              
Out[3]: 
⎛ 2            ⎞
⎝x  - 2, [1, 0]⎠
```

This PR gives:
```julia
In [29]: primitive_element([a, b])                                                                                                             
Out[29]: 
⎛ 2            ⎞
⎝x  - 2, [1, 0]⎠

In [30]: primitive_element([b, a])                                                                                                             
Out[30]: 
⎛ 2                  ⎞
⎝x  - 2⋅x - 1, [1, 0]⎠
```

I guess that's because the Groebner basis will be the same regardless of the input order:
```julia
In [2]: groebner([minpoly(a), minpoly(b)])                                                                                                     
Out[2]: 
             ⎛⎡ 2       2          ⎤                           ⎞
GroebnerBasis⎝⎣x  - 2, x  - 2⋅x - 1⎦, x, x, domain=ℤ, order=lex⎠

In [3]: groebner([minpoly(b), minpoly(a)])                                                                                                     
Out[3]: 
             ⎛⎡ 2       2          ⎤                           ⎞
GroebnerBasis⎝⎣x  - 2, x  - 2⋅x - 1⎦, x, x, domain=ℤ, order=lex⎠
```

The nondeterministic behaviour for `construct_domain` here comes from the fact that `construct_domain` uses a `set` for the extension elements before passing them to `primitive_element` which in this PR gives different results for different input orders:
https://github.com/sympy/sympy/blob/c08b0b37750fb915ed884a819e6ed208d4d9aa37/sympy/polys/constructor.py#L84

Potentially `primitive_element` could sort its inputs but also it could deconstruct the inputs to find the elementary extension elements like `[sqrt(3), I]`.

jksuom commented:
> Groebner basis will be the same regardless of the input order

I'm not sure that this would be true in general. The algorithm works with sets.

> Potentially primitive_element could sort its inputs but also it could deconstruct the inputs to find the elementary extension elements like [sqrt(3), I].

Maybe there could be a wrapper for this purpose. It would sort the inputs and then undo the changes before returning the results.

oscarbenjamin commented:
With this PR the output of `apart` can be non-deterministic. This is an example from #20423:
```python
In [1]: apart(1/(x**7-1), extension=cos(2*pi/7))                                                                                  
Out[1]: 
      ⎛     ⎛2⋅π⎞    ⎞           ⎛            2⎛2⋅π⎞    ⎞                      ⎛2⋅π⎞          2⎛2⋅π⎞                    
    2⋅⎜x⋅cos⎜───⎟ - 1⎟         2⋅⎜-x + 2⋅x⋅cos ⎜───⎟ - 1⎟          -x + 2⋅x⋅cos⎜───⎟ + 4⋅x⋅cos ⎜───⎟ + 2                
      ⎝     ⎝ 7 ⎠    ⎠           ⎝             ⎝ 7 ⎠    ⎠                      ⎝ 7 ⎠           ⎝ 7 ⎠               1    
───────────────────────── + ──────────────────────────────── - ───────────────────────────────────────────── + ─────────
  ⎛ 2          ⎛2⋅π⎞    ⎞     ⎛ 2          2⎛2⋅π⎞          ⎞     ⎛ 2              ⎛2⋅π⎞          2⎛2⋅π⎞    ⎞   7⋅(x - 1)
7⋅⎜x  - 2⋅x⋅cos⎜───⎟ + 1⎟   7⋅⎜x  - 4⋅x⋅cos ⎜───⎟ + 2⋅x + 1⎟   7⋅⎜x  - x + 2⋅x⋅cos⎜───⎟ + 4⋅x⋅cos ⎜───⎟ + 1⎟            
  ⎝            ⎝ 7 ⎠    ⎠     ⎝             ⎝ 7 ⎠          ⎠     ⎝                ⎝ 7 ⎠           ⎝ 7 ⎠    ⎠
```
In a new process:
```python
In [2]: apart(1/(x**7-1), extension=cos(2*pi/7))                                                                                  
Out[2]: 
                  ⎛2⋅π⎞          2⎛2⋅π⎞                    ⎛2⋅π⎞          2⎛2⋅π⎞               4⎛2⋅π⎞           3⎛2⋅π⎞          
      -x + 2⋅x⋅cos⎜───⎟ + 4⋅x⋅cos ⎜───⎟ + 2       - 6⋅x⋅cos⎜───⎟ - 8⋅x⋅cos ⎜───⎟ + x + 16⋅x⋅cos ⎜───⎟ + 16⋅x⋅cos ⎜───⎟ + 2   - 4
                  ⎝ 7 ⎠           ⎝ 7 ⎠                    ⎝ 7 ⎠           ⎝ 7 ⎠                ⎝ 7 ⎠            ⎝ 7 ⎠          
- ───────────────────────────────────────────── - ──────────────────────────────────────────────────────────────────────── + ───
    ⎛ 2              ⎛2⋅π⎞          2⎛2⋅π⎞    ⎞                         ⎛ 2          2⎛2⋅π⎞          ⎞                          
  7⋅⎜x  - x + 2⋅x⋅cos⎜───⎟ + 4⋅x⋅cos ⎜───⎟ + 1⎟                       7⋅⎜x  - 4⋅x⋅cos ⎜───⎟ + 2⋅x + 1⎟                          
    ⎝                ⎝ 7 ⎠           ⎝ 7 ⎠    ⎠                         ⎝             ⎝ 7 ⎠          ⎠                          

      ⎛2⋅π⎞          2⎛2⋅π⎞               4⎛2⋅π⎞           3⎛2⋅π⎞                
⋅x⋅cos⎜───⎟ - 4⋅x⋅cos ⎜───⎟ - x + 16⋅x⋅cos ⎜───⎟ + 16⋅x⋅cos ⎜───⎟ - 2            
      ⎝ 7 ⎠           ⎝ 7 ⎠                ⎝ 7 ⎠            ⎝ 7 ⎠           1    
───────────────────────────────────────────────────────────────────── + ─────────
                      ⎛ 2          ⎛2⋅π⎞    ⎞                           7⋅(x - 1)
                    7⋅⎜x  - 2⋅x⋅cos⎜───⎟ + 1⎟                                    
                      ⎝            ⎝ 7 ⎠    ⎠ 
```

czgdp1807 commented:
@jksuom Any thoughts on [the above comment](https://github.com/sympy/sympy/pull/20274#issuecomment-727231572)? And would you like to continue working on this PR? We will mark the PR, `Please Take Over` otherwise.

jksuom commented:
I have not been able to reproduce the above results. On my system, the results do not seem to vary:
```
In [1]: r1 = apart(1/(x**7-1), extension=cos(2*pi/7))                           

In [2]: r2 = apart(1/(x**7-1), extension=cos(2*pi/7))                           

In [3]: r3 = apart(1/(x**7-1), extension=cos(2*pi/7))                           

In [4]: r1 == r2                                                                
Out[4]: True

In [5]: r1 == r3                                                                
Out[5]: True
```
I think that the source of non-determinism must lie elsewhere as the primitive element algorithm is deterministic.

oscarbenjamin commented:
> the results do not seem to vary

You need to do each run in a separate process (exit isympy and restart it):
```console
$ cat t.py
from sympy import *
x = Symbol('x')
print(apart(1/(x**7-1), extension=cos(2*pi/7)))
$ python t.py 
2*(x*cos(2*pi/7) - 1)/(7*(x**2 - 2*x*cos(2*pi/7) + 1)) + 2*(-x + 2*x*cos(2*pi/7)**2 - 1)/(7*(x**2 - 4*x*cos(2*pi/7)**2 + 2*x + 1)) - (-x + 2*x*cos(2*pi/7) + 4*x*cos(2*pi/7)**2 + 2)/(7*(x**2 - x + 2*x*cos(2*pi/7) + 4*x*cos(2*pi/7)**2 + 1)) + 1/(7*(x - 1))
$ python t.py 
(-448*x*cos(2*pi/7)**5 - 96*x*cos(2*pi/7)**4 - 12*x*cos(2*pi/7) - x + 256*x*cos(2*pi/7)**8 + 20*x*cos(2*pi/7)**2 + 512*x*cos(2*pi/7)**7 + 128*x*cos(2*pi/7)**3 - 2)/(7*(x**2 - x + 2*x*cos(2*pi/7) + 4*x*cos(2*pi/7)**2 + 1)) - (-6*x*cos(2*pi/7) - 8*x*cos(2*pi/7)**2 + x + 16*x*cos(2*pi/7)**4 + 16*x*cos(2*pi/7)**3 + 2)/(7*(x**2 - 4*x*cos(2*pi/7)**2 + 2*x + 1)) - (-448*x*cos(2*pi/7)**5 - 112*x*cos(2*pi/7)**4 - 6*x*cos(2*pi/7) - x + 256*x*cos(2*pi/7)**8 + 28*x*cos(2*pi/7)**2 + 512*x*cos(2*pi/7)**7 + 112*x*cos(2*pi/7)**3 + 2)/(7*(x**2 - 2*x*cos(2*pi/7) + 1)) + 1/(7*(x - 1))
```
The non-determinism comes from the fact that the arguments passed to `primitive_element` are from a set and set order is non-deterministic across different process (because of hash-randomisation). The behaviour of hash randomisation also depends on the Python version (I'm testing with 3.8 here).

The difference this PR makes is that it means that the `primitive_elements` function here seems to give different results depending on the order of its inputs whereas the previous version based on the Groebner basis is unaffected by the order of its arguments.

jksuom commented:
> the arguments passed to primitive_element are from a set

Maybe arguments coming from a set should be ordered before the call.

jksuom commented:
This seems to give stable result:
```
--- a/sympy/polys/constructor.py
+++ b/sympy/polys/constructor.py
@@ -2,6 +2,7 @@
 
 
 from sympy.core import sympify
+from sympy.core.compatibility import ordered
 from sympy.core.evalf import pure_complex
 from sympy.polys.domains import ZZ, QQ, ZZ_I, QQ_I, EX
 from sympy.polys.domains.complexfield import ComplexField
@@ -102,7 +103,7 @@ def _construct_algebraic(coeffs, opt):
 
         result.append(coeff)
 
-    exts = list(exts)
+    exts = list(ordered(exts))
 
     g, span, H = primitive_element(exts, ex=True, polys=True)
     root = sum([ s*ext for s, ext in zip(span, exts) ])
```

jksuom committed:
Fix the arguments of primitive_element in _construct_algebraic.

The arguments are sorted to obtain reproducible results.

oscarbenjamin commented:
It would be good to have a test for `apart(1/(x**7-1), extension=cos(2*pi/7))` to see that the results are consistent.

oscarbenjamin commented on sympy/core/numbers.py:
```diff
@@ -2447,6 +2447,11 @@ class AlgebraicNumber(Expr):
+    # Optional alias symbol is not free.
+    # Actually, alias should be a Str, but some methods
+    # expect that it be an instance of Expr.
+    free_symbols = set()
```
Which methods are those? Maybe they should be changed.

jksuom replied:
There were some errors when I tried to use Str (I forget, what exactly) but I suppose that it would not be hard to fix them.

oscarbenjamin commented:
> It would be good to have a test for `apart(1/(x**7-1), extension=cos(2*pi/7))` to see that the results are consistent.

Perhaps also/instead a test that `construct_domain` gives the same result for items in different orders would be good.

oscarbenjamin commented:
Why do we need two different algorithms for `ex=True` and `ex=False`?

The approach here fails for a trivial case:
```
In [1]: primitive_element([1, 2])
Out[1]: (x - 1, [1, 0])

In [2]: primitive_element([1, 2], ex=True)
---------------------------------------------------------------------------
NotInvertible
```

jksuom commented:
I think that `ex=True` should be interpreted as "return also the expressions". The algorithms are essentially the same. That `NotInvertible` looks like a missing test someplace.

oscarbenjamin commented:
We can also improve performance with something like this to extract the elementary algebraic parts of the expression:
```diff
diff --git a/sympy/polys/constructor.py b/sympy/polys/constructor.py
index d6c5faa78e..22a4b3bf70 100644
--- a/sympy/polys/constructor.py
+++ b/sympy/polys/constructor.py
@@ -82,27 +82,24 @@ def _construct_algebraic(coeffs, opt):
     """We know that coefficients are algebraic so construct the extension. """
     from sympy.polys.numberfields import primitive_element
 
-    result, exts = [], set()
-
-    for coeff in coeffs:
-        if coeff.is_Rational:
-            coeff = (None, 0, QQ.from_sympy(coeff))
-        else:
-            a = coeff.as_coeff_add()[0]
-            coeff -= a
-
-            b = coeff.as_coeff_mul()[0]
-            coeff /= b
-
-            exts.add(coeff)
-
-            a = QQ.from_sympy(a)
-            b = QQ.from_sympy(b)
-
-            coeff = (coeff, b, a)
-
-        result.append(coeff)
+    exts = set()
+
+    def build_trees(args):
+        trees = []
+        for a in args:
+            if a.is_Rational:
+                tree = ('Q', QQ.from_sympy(a))
+            elif a.is_Add:
+                tree = ('+', build_trees(a.args))
+            elif a.is_Mul:
+                tree = ('*', build_trees(a.args))
+            else:
+                tree = ('e', a)
+                exts.add(a)
+            trees.append(tree)
+        return trees
 
+    trees = build_trees(coeffs)
     exts = list(ordered(exts))
 
     g, span, H = primitive_element(exts, ex=True, polys=True)
@@ -110,13 +107,27 @@ def _construct_algebraic(coeffs, opt):
 
     domain, g = QQ.algebraic_field((g, root)), g.rep.rep
 
-    for i, (coeff, a, b) in enumerate(result):
-        if coeff is not None:
-            coeff = a*domain.dtype.from_list(H[exts.index(coeff)], g, QQ) + b
+    exts_dom = [domain.dtype.from_list(h, g, QQ) for h in H]
+    exts_map = dict(zip(exts, exts_dom))
+
+    def convert_tree(tree):
+        op, args = tree
+        if op == 'Q':
+            return domain.dtype.from_list([args], g, QQ)
+        elif op == '+':
+            return sum((convert_tree(a) for a in args), domain.zero)
+        elif op == '*':
+            # return prod(convert(a) for a in args)
+            t = convert_tree(args[0])
+            for a in args[1:]:
+                t *= convert_tree(a)
+            return t
+        elif op == 'e':
+            return exts_map[args]
         else:
-            coeff = domain.dtype.from_list([b], g, QQ)
+            raise RuntimeError
 
-        result[i] = coeff
+    result = [convert_tree(tree) for tree in trees]
 
     return domain, result
 
diff --git a/sympy/polys/rings.py b/sympy/polys/rings.py
index 3585e8f0bf..9d7fe18871 100644
--- a/sympy/polys/rings.py
+++ b/sympy/polys/rings.py
@@ -163,10 +163,12 @@ def sring(exprs, *symbols, **options):
     reps, opt = _parallel_dict_from_expr(exprs, opt)
 
     if opt.domain is None:
-        # NOTE: this is inefficient because construct_domain() automatically
-        # performs conversion to the target domain. It shouldn't do this.
         coeffs = sum([ list(rep.values()) for rep in reps ], [])
-        opt.domain, _ = construct_domain(coeffs, opt=opt)
+
+        opt.domain, coeffs_dom = construct_domain(coeffs, opt=opt)
+
+        coeff_map = dict(zip(coeffs, coeffs_dom))
+        reps = [{m: coeff_map[c] for m, c in rep.items()} for rep in reps]
 
     _ring = PolyRing(opt.gens, opt.domain, opt.order)
     polys = list(map(_ring.from_dict, reps))
```

I tested performance with:
```python
from sympy import *

exts = [sqrt(prime(n)) for n in range(1, 10)]

nexts = 2

for n in range(1, 10):
    M = randMatrix(n, n + 1)
    for ext in exts[:nexts]:
        M += ext * randMatrix(n, n + 1)
    syms = symbols('x:%d' % n)
    eqs = list(M * Matrix(syms + (1,)))

    import time
    start = time.time()
    linsolve(eqs, syms)
    end = time.time()
    print('n =', n, ':', '%2g secs' % (end - start))
```
On master that gives:
```console
$ python time.py 
n = 1 : 3.18847 secs
^C
```

With this PR that gives:
```console
$ python time.py 
n = 1 : 0.229376 secs
n = 2 : 0.747598 secs
n = 3 : 1.62068 secs
n = 4 : 2.94355 secs
n = 5 : 3.71783 secs
n = 6 : 6.04225 secs
^C
```

With this PR and the diff above it gives:
```console
$ python time.py 
n = 1 : 0.0299489 secs
n = 2 : 0.0353289 secs
n = 3 : 0.0492551 secs
n = 4 : 0.0669062 secs
n = 5 : 0.0878379 secs
n = 6 : 0.127864 secs
n = 7 : 0.154448 secs
n = 8 : 0.201635 secs
n = 9 : 0.256783 secs
```

@jksuom do you mind if I push that on here?

jksuom commented:
Those changes look good (the timings almost too good to be true). Please push them. (That will save me the trouble:)

oscarbenjamin committed:
perf(polys): speed up construction of algebraic domains

Filter out elementary non-rationals before calling primitive element to
reduce the size of the inputs so e.g.

   construct_domain([sqrt(2) + 3*sqrt(3), -sqrt(2) + 2, ...])

will only call primitive_element([sqrt(2), sqrt(3)]) rather than passing
every element.

Also constructing a ring is made faster by using the converted elements
from construct_domain

oscarbenjamin commented:
I've just pushed that on and changed one test:
```diff
diff --git a/sympy/polys/tests/test_rings.py b/sympy/polys/tests/test_rings.py
index 274ac4c702..2d2666c3b1 100644
--- a/sympy/polys/tests/test_rings.py
+++ b/sympy/polys/tests/test_rings.py
@@ -171,7 +171,7 @@ def test_sring():
 
     r = sqrt(2) - sqrt(3)
     R, a = sring(r, extension=True)
-    assert R.domain == QQ.algebraic_field(r)
+    assert R.domain == QQ.algebraic_field(sqrt(2) + sqrt(3))
     assert R.gens == ()
     assert a == R.domain.from_sympy(r)
```
Here the extension element gets normalised by sring.

> the timings almost too good to be true

If you increase `next` the number of extensions then it slows down quite quickly:
```console
$ python time.py    # next=3
n = 1 : 0.080785 secs
n = 2 : 0.0825591 secs
n = 3 : 0.115191 secs
n = 4 : 0.149894 secs
n = 5 : 0.266975 secs
n = 6 : 0.43802 secs
n = 7 : 0.432203 secs
n = 8 : 0.587976 secs
n = 9 : 0.792337 secs

$ python time.py    # next=4
n = 1 : 0.74924 secs
n = 2 : 1.6284 secs
n = 3 : 2.11424 secs
n = 4 : 2.90014 secs
n = 5 : 3.78933 secs
n = 6 : 4.93005 secs
^CTraceback (most recent call last):
  File "time.py", line 16, in <module>
    linsolve(eqs, syms)
  File "/Users/enojb/current/sympy/sympy/sympy/solvers/solveset.py", line 2645, in linsolve
    sol = solve_lin_sys(eqs, ring, _raw=False)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/solvers.py", line 238, in solve_lin_sys
    tresult = {to_sympy(sym): to_sympy(val) for sym, val in result.items()}
  File "/Users/enojb/current/sympy/sympy/sympy/polys/solvers.py", line 238, in <dictcomp>
    tresult = {to_sympy(sym): to_sympy(val) for sym, val in result.items()}
  File "/Users/enojb/current/sympy/sympy/sympy/polys/solvers.py", line 236, in to_sympy
    return ring.domain.to_sympy(x)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/domains/algebraicfield.py", line 64, in to_sympy
    return AlgebraicNumber(self.ext, a).as_expr()
  File "/Users/enojb/current/sympy/sympy/sympy/core/numbers.py", line 2532, in as_expr
    return self.as_poly(x or self.root).as_expr().expand()
  File "/Users/enojb/current/sympy/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 3579, in expand
    expr, hit = Expr._expand_hint(expr, hint, deep=deep, **hints)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 3512, in _expand_hint
    arg, arghit = Expr._expand_hint(arg, hint, **hints)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 3512, in _expand_hint
    arg, arghit = Expr._expand_hint(arg, hint, **hints)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 3520, in _expand_hint
    newexpr = getattr(expr, hint)(**hints)
  File "/Users/enojb/current/sympy/sympy/sympy/core/power.py", line 1162, in _eval_expand_multinomial
    return basic_from_dict(expansion_dict, *p)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polyutils.py", line 382, in expr_from_dict
    term.append(Pow(g, m))
  File "/Users/enojb/current/sympy/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 120, in __hash__
    h = self._mhash
KeyboardInterrupt
```
The slow part there seems to converting the solution back to Expr at the end.

oscarbenjamin committed:
perf(polys): speed up conversion from algebraic domains

oscarbenjamin commented:
I've just pushed a commit to speed up converting from algebraic fields:
```console
$ python time.py    # next=4
n = 1 : 0.23641 secs
n = 2 : 0.152647 secs
n = 3 : 0.268682 secs
n = 4 : 0.468488 secs
n = 5 : 0.848626 secs
n = 6 : 1.37821 secs
n = 7 : 2.22053 secs
n = 8 : 3.50289 secs
n = 9 : 5.1078 secs

$ python time.py   # nexts=5
n = 1 : 1.3422 secs
n = 2 : 1.90083 secs
n = 3 : 4.09537 secs
n = 4 : 8.02378 secs
n = 5 : 15.5078 secs
n = 6 : 26.7734 secs
^CTraceback (most recent call last):
  File "time.py", line 17, in <module>
    linsolve(eqs, syms)
  File "/Users/enojb/current/sympy/sympy/sympy/solvers/solveset.py", line 2645, in linsolve
    sol = solve_lin_sys(eqs, ring, _raw=False)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/solvers.py", line 227, in solve_lin_sys
    result = _solve_lin_sys(eqs_coeffs, eqs_rhs, ring)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/solvers.py", line 305, in _solve_lin_sys
    subsol = _solve_lin_sys_component(subsystem[0], subsystem[1], ring)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/solvers.py", line 353, in _solve_lin_sys_component
    echelon, pivots = matrix.rref()
  File "/Users/enojb/current/sympy/sympy/sympy/polys/domainmatrix.py", line 609, in rref
    rref_ddm, pivots = self.rep.rref()
  File "/Users/enojb/current/sympy/sympy/sympy/polys/domainmatrix.py", line 141, in rref
    pivots = ddm_irref(b)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/domainmatrix.py", line 257, in ddm_irref
    ai[l] /= aij # ai[j] = one
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polyclasses.py", line 1723, in __truediv__
    return f.quo(g)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polyclasses.py", line 1646, in quo
    return per(dup_rem(dup_mul(F, dup_invert(G, mod, dom), dom), mod, dom))
  File "/Users/enojb/current/sympy/sympy/sympy/polys/euclidtools.py", line 170, in dup_invert
    s, h = dup_half_gcdex(f, g, K)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/euclidtools.py", line 78, in dup_half_gcdex
    a, b = b, dup_sub_mul(a, q, b, K)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/densearith.py", line 715, in dup_sub_mul
    return dup_sub(f, dup_mul(g, h, K), K)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/densearith.py", line 767, in dup_mul
    coeff += f[j]*g[i - j]
KeyboardInterrupt
```
Now it seems that the time is mostly spent in the actual row reduction step of the solver.

oscarbenjamin commented:
I can make it about faster with this:
```diff
diff --git a/sympy/polys/domainmatrix.py b/sympy/polys/domainmatrix.py
index 2bf8cce3c6..373ac4483a 100644
--- a/sympy/polys/domainmatrix.py
+++ b/sympy/polys/domainmatrix.py
@@ -253,8 +253,9 @@ def ddm_irref(a):
 
         # normalise row
         ai = a[i]
+        aijinv = aij**-1
         for l in range(j, n):
-            ai[l] /= aij # ai[j] = one
+            ai[l] *= aijinv # ai[j] = one
 
         # eliminate above and below to the right
         for k, ak in enumerate(a):
```
That gives:
```console
$ python time.py     # next=5
n = 1 : 1.35528 secs
n = 2 : 1.22998 secs
n = 3 : 2.32967 secs
n = 4 : 4.24498 secs
n = 5 : 8.28188 secs
n = 6 : 12.6741 secs
n = 7 : 19.2178 secs
^CTraceback (most recent call last):
  File "time.py", line 17, in <module>
    linsolve(eqs, syms)
  File "/Users/enojb/current/sympy/sympy/sympy/solvers/solveset.py", line 2645, in linsolve
    sol = solve_lin_sys(eqs, ring, _raw=False)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/solvers.py", line 227, in solve_lin_sys
    result = _solve_lin_sys(eqs_coeffs, eqs_rhs, ring)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/solvers.py", line 305, in _solve_lin_sys
    subsol = _solve_lin_sys_component(subsystem[0], subsystem[1], ring)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/solvers.py", line 353, in _solve_lin_sys_component
    echelon, pivots = matrix.rref()
  File "/Users/enojb/current/sympy/sympy/sympy/polys/domainmatrix.py", line 610, in rref
    rref_ddm, pivots = self.rep.rref()
  File "/Users/enojb/current/sympy/sympy/sympy/polys/domainmatrix.py", line 141, in rref
    pivots = ddm_irref(b)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/domainmatrix.py", line 256, in ddm_irref
    aijinv = aij**-1
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polyclasses.py", line 1713, in __pow__
    return f.pow(n)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polyclasses.py", line 1622, in pow
    F, n = dup_invert(f.rep, f.mod, f.dom), -n
  File "/Users/enojb/current/sympy/sympy/sympy/polys/euclidtools.py", line 170, in dup_invert
    s, h = dup_half_gcdex(f, g, K)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/euclidtools.py", line 80, in dup_half_gcdex
    a = dup_quo_ground(a, dup_LC(f, K), K)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/densearith.py", line 321, in dup_quo_ground
    return [ K.quo(cf, c) for cf in f ]
  File "/Users/enojb/current/sympy/sympy/sympy/polys/densearith.py", line 321, in <listcomp>
    return [ K.quo(cf, c) for cf in f ]
  File "/Users/enojb/current/sympy/sympy/sympy/polys/domains/gmpyrationalfield.py", line 76, in quo
    return GMPYRational(a) / GMPYRational(b)
KeyboardInterrupt
```
The bulk of the time for smaller systems is still spent computing that inverse `aijinv ** -1` even if we only do it once per row. I wonder if there's a way to speed that up.

oscarbenjamin committed:
perf(polys): only compute the inverse of the pivot once

oscarbenjamin committed:
fix(polys): use elements of a consistent domain

Ensure that the back-substitution step of solve_lin_sys performs all
arithmetic with PolyElement rather than mixing in elements of the
ground domain.

oscarbenjamin commented:
@jksuom what do you think of this now?

I think it's good to merge if you're happy with it.

jksuom commented:
I also think that it is good to merge.

oscarbenjamin commented:
Greaat. Thanks!
